### PR TITLE
metal : implement q5_0 and q5_1 kernels

### LINKED ISSUE #3504: metal : add Q5_0 and Q5_1 support
The implementation for `Q5_0` and `Q5_1` is still missing in the Metal backend. We should add it

https://github.com/ggerganov/llama.cpp/blob/5ab6c2132aad2354092a26c096cc5c8f55801141/ggml-metal.m#L962-L974

https://github.com/ggerganov/llama.cpp/blob/5ab6c2132aad2354092a26c096cc5c8f55801141/ggml-metal.m#L996-L1003

### PR DESCRIPTION
**Pull request by jhen0409**
Implement dequantize functions & mul_mv kernels for #3504.

```bash
./llama-bench -m models/7B/ggml-model-llama-2-7b-q4_0.gguf \
  -m models/7B/ggml-model-llama-2-7b-q4_1.gguf \
  -m models/7B/ggml-model-llama-2-7b-q5_0.gguf \
  -m models/7B/ggml-model-llama-2-7b-q5_1.gguf \
  -m models/7B/ggml-model-llama-2-7b-q8_0.gguf -t 4
```

M1 Max (32c GPU):

| model                          |       size |     params | backend    | ngl |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 |          4 | pp 512     |   514.94 ± 10.77 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 |          4 | tg 128     |     61.87 ± 0.69 |
| llama 7B mostly Q4_1           |   3.95 GiB |     6.74 B | Metal      |  99 |          4 | pp 512     |   515.85 ± 10.63 |
| llama 7B mostly Q4_1           |   3.95 GiB |     6.74 B | Metal      |  99 |          4 | tg 128     |     57.82 ± 0.57 |
| llama 7B mostly Q5_0           |   4.33 GiB |     6.74 B | Metal      |  99 |          4 | pp 512     |    450.06 ± 7.63 |
| llama 7B mostly Q5_0           |   4.33 GiB |     6.74 B | Metal      |  99 |          4 | tg 128     |     42.72 ± 0.32 |
| llama 7B mostly Q5_1           |   4.72 GiB |     6.74 B | Metal      |  99 |          4 | pp 512     |    440.58 ± 3.19 |
| llama 7B mostly Q5_1           |   4.72 GiB |     6.74 B | Metal      |  99 |          4 | tg 128     |     42.23 ± 0.39 |
| llama 7B mostly Q8_0           |   6.67 GiB |     6.74 B | Metal      |  99 |          4 | pp 512     |    522.44 ± 9.21 |
| llama 7B mostly Q8_0           |   6.67 GiB |     6.74 B | Metal      |  99 |          4 | tg 128     |     40.29 ± 0.46 |

M2 (10c GPU):

| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |   1 | pp 512     |    178.89 ± 0.08 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |   1 | tg 128     |     20.85 ± 0.07 |
| llama 7B mostly Q4_1           |   3.95 GiB |     6.74 B | Metal      |   1 | pp 512     |    179.21 ± 0.07 |
| llama 7B mostly Q4_1           |   3.95 GiB |     6.74 B | Metal      |   1 | tg 128     |     18.97 ± 0.06 |
| llama 7B mostly Q5_0           |   4.33 GiB |     6.74 B | Metal      |   1 | pp 512     |    153.31 ± 0.06 |
| llama 7B mostly Q5_0           |   4.33 GiB |     6.74 B | Metal      |   1 | tg 128     |     11.04 ± 2.70 |
| llama 7B mostly Q5_1           |   4.72 GiB |     6.74 B | Metal      |   1 | pp 512     |    115.67 ± 9.05 |
| llama 7B mostly Q5_1           |   4.72 GiB |     6.74 B | Metal      |   1 | tg 128     |     10.87 ± 0.28 |
| llama 7B mostly Q8_0           |   6.67 GiB |     6.74 B | Metal      |   1 | pp 512     |   154.55 ± 19.90 |
| llama 7B mostly Q8_0           |   6.67 GiB |     6.74 B | Metal      |   1 | tg 128     |     10.44 ± 0.55 |

> I experience issue like downclocking on my M2 Macbook Air, so this may not be a very accurate result.

build: ad800e8 (1358)

## TODO
- [x] dequantize_q5_0
- [x] kernel_mul_mv_q5_0_f32
- [x] dequantize_q5_1
- [x] kernel_mul_mv_q5_1_f32
- [x] bench result

### CODE REVIEW

File: ggml-metal.metal [hunk 1]
```diff
@@ -428,6 +442,32 @@ inline float block_q_n_dot_y(device const block_q4_1 * qb_curr, float sumy, thre
...
+    device const uint16_t * qs = ((device const uint16_t *)qb_curr + 3 + il/2);
+    const uint32_t qh = *((device const uint32_t *)qb_curr->qh);
+    for (int i = 0; i < 8; i+=2) {
+        acc[0] += yl[i + 0] * ((qs[i / 2] & 0x000F) | ((qh >> (i+0+il        ) << 4 ) & 0x0010))
+                + yl[i + 1] * ((qs[i / 2] & 0x0F00) | ((qh >> (i+1+il        ) << 12) & 0x1000));
+        acc[1] += yl[i + 8] * ((qs[i / 2] & 0x00F0) | ((qh >> (i+0+il+QK5_0/2) << 4 ) & 0x0010))
+                + yl[i + 9] * ((qs[i / 2] & 0xF000) | ((qh >> (i+1+il+QK5_0/2) << 12) & 0x1000));
+    }
+    return d * (sumy * -16.f + acc[0] + acc[1]);
+}
```
Comment (jhen0409) [inline ggml-metal.metal:461]: Need to correct it, I'll keep check later. Any help will be appreciated.

File: ggml-metal.metal [hunk 2]
```diff
@@ -428,6 +442,32 @@ inline float block_q_n_dot_y(device const block_q4_1 * qb_curr, float sumy, thre
...
 }
 
+// function for calculate inner product between half a q5_0 block and 16 floats (yl), sumy is SUM(yl[i])
+// il indicates where the q5 quants begin (0 or QK5_0/4)
+// we assume that the yl's have been multiplied with the appropriate scale factor
+// that corresponds to the missing bit shifts (1, 1/16, 1/256, 1/4096)
+inline float block_q_n_dot_y(device const block_q5_0 * qb_curr, float sumy, thread float * yl, int il) {
+    float d = qb_curr->d;
+    float2 acc = 0.f;
+    device const uint16_t * qs = ((device const uint16_t *)qb_curr + 3 + il/2);
```
Comment (ggerganov) [inline ggml-metal.metal:452]:
```suggestion
    device const uint16_t * qs = ((device const uint16_t *)qb_curr + 1 + il/2);
```

Comment (jhen0409) [reply ggml-metal.metal:452]: +3 should be correct because it skip 4 uint8_t (qh). Am I wrong?

Comment (ggerganov) [reply ggml-metal.metal:452]: Yes, sorry I missed that

**Comment by jhen0409** [comment]
It works, but a little bit slow. In addition to the reason of 5-bit extraction, there should be some small optimization for operations.

**Comment by ggerganov** [comment]
Thank you, I think even if the performance is not there we can improve it from `master`.
Can you verify that the perplexity produces reasonable numbers - both with `-b 512` and with `-b 1`?

**Comment by jhen0409** [comment]
```
./perplexity -m ./models/ggml-model-llama-2-7b-q5_0.gguf -f ./wikitext-2-raw/wiki.test.raw -t 4 -ngl 1 -b 512

[1]4.1671,[2]4.7427,[3]5.3947,[4]5.9889,[5]6.1096,[6]6.0134,[7]6.1760,[8]6.2610,[9]6.5658,[10]6.7533

./perplexity -m ./models/ggml-model-llama-2-7b-q5_0.gguf -f ./wikitext-2-raw/wiki.test.raw -t 4 -ngl 1 -b 1

[1]4.1672,[2]4.7428,[3]5.3947,[4]5.9890,[5]6.1096,[6]6.0134,[7]6.1761,[8]6.2611,[9]6.5659,[10]6.7534

./perplexity -m ./models/ggml-model-llama-2-7b-q5_1.gguf -f ./wikitext-2-raw/wiki.test.raw -t 4 -ngl 1 -b 512

[1]4.2048,[2]4.7178,[3]5.3836,[4]5.9347,[5]6.0646,[6]5.9762,[7]6.1550,[8]6.2462,[9]6.5789,[10]6.7625

./perplexity -m ./models/ggml-model-llama-2-7b-q5_1.gguf -f ./wikitext-2-raw/wiki.test.raw -t 4 -ngl 1 -b 1

[1]4.2047,[2]4.7177,[3]5.3835,[4]5.9346,[5]6.0644,[6]5.9761,[7]6.1548,[8]6.2461,[9]6.5788,[10]6.7625
```

Looks good.

**Review by ggerganov — APPROVED** [review]
Here is a bench on M2 Ultra:

| model                          |       size |     params | backend    | ngl |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |   1 |          4 | pp 512     |   1235.45 ± 0.51 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |   1 |          4 | tg 128     |     98.16 ± 0.06 |
| llama 7B mostly Q4_1           |   3.95 GiB |     6.74 B | Metal      |   1 |          4 | pp 512     |   1236.97 ± 1.04 |
| llama 7B mostly Q4_1           |   3.95 GiB |     6.74 B | Metal      |   1 |          4 | tg 128     |     90.27 ± 0.04 |
| llama 7B mostly Q5_0           |   4.33 GiB |     6.74 B | Metal      |   1 |          4 | pp 512     |   1067.33 ± 0.74 |
| llama 7B mostly Q5_0           |   4.33 GiB |     6.74 B | Metal      |   1 |          4 | tg 128     |     78.74 ± 0.07 |
| llama 7B mostly Q5_1           |   4.72 GiB |     6.74 B | Metal      |   1 |          4 | pp 512     |   1068.07 ± 1.02 |
| llama 7B mostly Q5_1           |   4.72 GiB |     6.74 B | Metal      |   1 |          4 | tg 128     |     75.76 ± 0.19 |

**Pull request merged**